index/scorch/persister.go

```diff
@@ -35,16 +37,27 @@ import (
 
 var DefaultChunkFactor uint32 = 1024
 
-// Arbitrary number, need to make it configurable.
-// Lower values like 10/making persister really slow
-// doesn't work well as it is creating more files to
-// persist for in next persist iteration and spikes the # FDs.
-// Ideal value should let persister also proceed at
-// an optimum pace so that the merger can skip
-// many intermediate snapshots.
-// This needs to be based on empirical data.
-// TODO - may need to revisit this approach/value.
-var epochDistance = uint64(5)
+var DefaultWorkLoopWait int = 2000 // ms
+
+var MaxWorkLoopWait int = 5000 // ms
+
+// ErrMaxWorkLoopWait is returned when the injected wait
+// for persistence loop is more than the maximum limit
+var ErrMaxWorkLoopWait = errors.New("WorkLoopWait exceeds the max wait limit")
+
+var DefaultPauseOnFileCount int = 1000
+
+type persisterOptions struct {
+	// WorkLoopWait controls the wait/delay injected into
+	// persistence workloop to improve the chances for
+	// a healthier and heavier in-memory merging
+	WorkLoopWait int
+
+	// PauseOnFileCount threshold mandates the persister to wait
+	// until the merger compacts the files on disk within this
+	// threshold
+	PauseOnFileCount int
```

Comment:
One thought was PauseOnFileCount might need a rename.

And, related, an edge case popped to mind... imagine there's a whole bunch of N files on disk (because there's a lot of data, perhaps due to the app wanting to ingest a lot of data)... but those files are already in "perfect staircase" formation.

In that case, the merger wakes up and thinks "cool, there's nothing to merge; I'm done".

But, if N is > this threshold option, then the persister would be incorrectly blocked.

So, I'm thinking the option that you'd really want to specify would be more like "pause the persister if the # of _mergable_ files goes above this threshold"?

Reply:
hm, thats a interesting angle.
But as a measure to limit the FD count, irrespective of the potential for further merge operations - we may still want to block persister as the number of FD's/files are over threshold at this point and by blocking persister the file count eventually comes under threshold either due to completion of concurrent queries or due to merger progress.?

Reply:
right -- i think it's perhaps because i misunderstood what PauseOnFileCount meant.

I initially took it to mean (when I wrote the above comment)... "pause the persister if the file count is greater than X"... but it actually means... "slow down or sleep the persister for a finite amount of time as long as the file count is < X"

Reply:
hi @steveyen , 
MaxNumberOfFiles / MaxFilesOnDisk or any other thoughts on the naming?